monai/data/dataset.py

```diff
@@ -921,19 +905,16 @@ def _transform(self, index: int):
             data = self._cache[cache_index] = self._load_cache_item(cache_index)
 
         # load data from cache and execute from the first random transform
-        start_run = False
         if not isinstance(self.transform, Compose):
             raise ValueError("transform must be an instance of monai.transforms.Compose.")
-        for _transform in self.transform.transforms:
-            if start_run or isinstance(_transform, RandomizableTrait) or not isinstance(_transform, Transform):
-                # only need to deep copy data on first non-deterministic transform
-                if not start_run:
-                    start_run = True
-                    if self.copy_cache:
-                        data = deepcopy(data)
-                data = self.transform.evaluate_with_overrides(data, _transform)
-                data = apply_transform(_transform, data)
-        data = self.transform.evaluate_with_overrides(data, None)
+
+        first_random = self.transform.get_index_of_first(
+            lambda t: isinstance(t, RandomizableTrait) or not isinstance(t, Transform)
+        )
+        if first_random is not None:
```

Comment:
I don't think that that is the case. If get_index_of_first is None, it means that the whole pipeline was processed in the caching step and so the only thing that needs to happen is the returning of the data. We aren't persisting data to the cache with pending transforms, so there cannot be pending transforms on the just loaded data

Reply:
ideally we should provide an option so that cache vs non-cache will achieve exactly the sample outcome.

Reply:
Not at all. That would create so many potential problems upstream and downstream. Pending transforms are by their nature easy and lightweight to calculate. As such, recalculating them is not a problem. By contrast, an executed transform is heavy weight and destructive. If you are going to the problem of caching data, then you have already committed to that expensive step. If you don't want to execute your pipeline, don't cache. If you do, then execute fully first. We don't provide people a way to do that either as compose guarantees full execution of a whole or partial pipeline

Reply:
If for some reason we decided that we wanted to support that, I'd definitely want to already refactored the caching mechanisms to work through structural transforms as discussed in the full design

Reply:
The only way that they can have the same sample outcome is if the caching step never needed to happen in the first place. For such a pipeline, you simply wouldn't cache in the first place

Reply:
also enforcing resampling when caching introduces interpolation errors when compared with the non-caching pipeline... I still think we should make it optional

Reply:
I think I'd need to see an example to be convinced. Either you cache before you do the multi-sampling or you cache after you do the multi-sampling, or you don't cache at all. Under what circumstance do you think there would be higher memory usage?

Reply:
Even then. I don't think this is a good feature to commit to at present. Even if it turns out to have a point, I would absolutely not feel comfortable supporting it without first thinking through all of the ramifications. Features like this need very careful design consideration, and it certainly hasn't had that.

Reply:
> I assume you are mainly talking about computational time, in terms of memory efficiency, keeping the pending operations have advantages when multiple samples sharing the same volume reference.

What is the example that reduces memory efficiency?

Reply:
In any case, we already have inefficiencies in the pipeline that we need the full design to iron out. I am happy to discuss this from a design perspective when we have the next design review of lazy resampling. Now is absolutely not the time to commit to this idea from a design standpoint. I certainly wouldn't want to support it by accident

Reply:
thanks, for example:

1 Loadimage: spatial size (32, 32)
2 Resize(output_size=(2000, 2000))
3 RandCrop(output_size=(40, 40))
4 RandRotate(angle=pi/4, keep_size=True)

lazy without cache, the resample is at step 4, with cache, it's at step 2 and 4. we can discuss in the meeting to agree on the default behaviour...

Reply:
This is an example of a pipeline where you do not benefit from caching and shouldn't cache.

I am not at all convinced that we should support pending transforms, and that a decision to do so should only follow a careful examination of the consequences of doing so. There is no time available for that for 1.2

Reply:
> This is an example of a pipeline where you do not benefit from caching and shouldn't cache. Not to mention a pipeline that wouldn't make sense to make in the first place.
> 

that I disagree... LoadImage + Spacing (Resizing) is used everywhere and we use cachedataset to hide the IO latency

Reply:
Yes, so there you cache. That causes the resample to happen. There isn't any point in loading and specifying the spacing, only to cache the transform with the pending spacing. You might as well just run the thing uncached.

Reply:
We absolutely could revisit the nature and flexibility of caching. I think that there is a lot that can be done when you phrase caching in terms of caching transforms, for example, because you can specify all kinds of clever behaviour in a way that is independent of the actual caching mechanism. We are close to a 1.2 code freeze, so this probably isn't the time for that.

I'll say it again; we don't understand the ramifications of doing it. I do not think it should be supported in any fashion without very careful examination of the consequences of that design change.

Reply:
> 

LoadImage, RandIdentity, Spacing / Resizing does this for you, no?

Reply:
sure, sounds good, as long as we are aware of the use cases of caching and documentation is clear then I'm fine with these as well.. cc @ericspod  @rijobro  @Nic-Ma @KumoLiu @yiheng-wang-nv

Reply:
I see @atbenmurray 's point that we don't have a clear design and understanding of this sort of caching here so we should keep things simple.